src/libs/Network/SequentialQueue.js

```diff
@@ -24,20 +25,42 @@ let currentRequest = null;
  * @returns {Promise}
  */
 function process() {
-    const persistedRequests = PersistedRequests.getAll();
+    const persistedRequests = [...PersistedRequests.getAll()];
 
-    // This sanity check is also a recursion exit point
-    if (NetworkStore.isOffline() || _.isEmpty(persistedRequests)) {
+    // If we have no persisted requests or we are offline we don't want to make any requests so we return early
+    if (_.isEmpty(persistedRequests) || NetworkStore.isOffline()) {
+        isSequentialQueueRunning = false;
         return Promise.resolve();
     }
 
-    const task = _.reduce(persistedRequests, (previousRequest, request) => previousRequest.then(() => {
-        currentRequest = Request.processWithMiddleware(request, true);
-        return currentRequest;
-    }), Promise.resolve());
+    isSequentialQueueRunning = true;
 
-    // Do a recursive call in case the queue is not empty after processing the current batch
-    return task.then(process);
+    // Get the first request in the queue and process it
+    const requestToProcess = persistedRequests.shift();
+
+    // Set the current request to a promise awaiting its processing
+    currentRequest = Request.processWithMiddleware(requestToProcess, true).then(() => {
+        // If the request is successful we want to:
+        // - Remove it from the queue
+        // - Clear any wait time we may have added if it failed before
+        // - Call process again to process the other requests in the queue
+        PersistedRequests.remove(requestToProcess);
+        RequestThrottle.clear();
+        return process();
+    }).catch((error) => {
+        // If the request was cancelled we don't want to retry it, so remove it from the queue and move on to the next request
```

Comment:
Why not?

Reply:
When we sign out we cancel all requests that are in progress with fetch. We don't want to merge in Onyx data that is sent from another account. We call [MainQueue.clear()](https://github.com/Expensify/App/blob/5c2875753752d1a7377a8687d1a78e5dd0f64136/src/libs/actions/SignInRedirect.js#L61) in redirectToSignIn.

It's a bit confusing because although we're clearing the MainQueue we call HttpUtils.cancelPendingRequests() in the process. That calls abort on the cancellation controller that is used by fetch within [processHTTPRequest](https://github.com/Expensify/App/blob/5c2875753752d1a7377a8687d1a78e5dd0f64136/src/libs/HttpUtils.js#L45-L51). Both the MainQueue and SequentialQueue use processHTTPRequest.

Reply:
Yeah good idea, updated!